### DISCUSSION TITLE
3.5.0 Ambiguous error trace

### DISCUSSION THREAD

**Discussion by JeremyJaydan** [alpinejs/alpine#2312] [1. Help] [answered]
I ran into an error that was hard to debug with the #2027 update.

With a previous Alpine version (3.4.2), I was able to view the root cause of an error in my code by clicking the "VM link"
![](https://user-images.githubusercontent.com/15324232/140200850-c79ce850-9964-457e-891c-45a83ee41531.png)
![](https://user-images.githubusercontent.com/15324232/140200897-08681776-6dd2-4907-a36d-1a9613932454.png)

however with version 3.5.0, there's an alpine script link:
![](https://user-images.githubusercontent.com/15324232/140201197-faca50b1-aa4e-4e43-807e-a489c194bd3f.png)
![](https://user-images.githubusercontent.com/15324232/140201310-2f2bb19e-033c-4c68-bd21-9743097e9e1d.png)

Is there a way to see the root cause in 3.5.0 or anything I'm missing I should use to debug easier?

thanks!

**Comment by danddanddand** [comment, marked as answer]
I think it should be making it easier to debug because it should be logging element and expression this occurred. This might not be case we tested (unescaped single quote) so I'll play with it. If you could replicate in codepen might help.

**Reply by JeremyJaydan** [reply]
sure thing, here's a pen: https://codepen.io/JeremyJaydan/full/ZEJvmXd

**Reply by danddanddand** [reply]
Thanks for the example. I think the error handling is working how it should in this case. Basically, before 3.5.0 there was no error handling when expressions were evaluated, so the first example you shared it was an x-data failing and giving reasonable way for you to trace it because it was showing "snippet" of your code throwing exception but it was still giving you stack trace from within Alpine. This would usually stop Alpine from continuing to execute other expressions and if your failing x-data was component with code that was duplicated on page (very common with Alpine) it could be hard to trace which particular element/expression was the problem element.

In 3.5.0+, when the expression is evaluated and fails, it's being moved off main execution thread, so alpine can more or less ignore it and continue. You especially want this behaviour in production environment so perhaps an unimportant component in footer doesn't cause Alpine to stop evaluating everything else on page. You should be able to see log of problem element/expression which provides same information but as in your example, may be slightly harder to interpret. In Chrome DevTools, you can right click on that element now and use "Reveal in Element Panel" to see more context. 

Perhaps you can experiment with @HugoDF's Alpine Dev Tools (https://github.com/alpine-collective/alpinejs-devtools) and see if it helps with debugging. For our dev team, the changes in 3.5.0 made it much easier to debug (often the only way to debug) so please give it a week or two and report back but I think it's working as intended.

**Reply by HugoDF** [reply]
The devtools doesn't support the 3.5.0 error handling but I'll have a look at supporting it.

**Reply by JeremyJaydan** [reply]
> Thanks for the example. I think the error handling is working how it should in this case. Basically, before 3.5.0 there was no error handling when expressions were evaluated, so the first example you shared it was an x-data failing and giving reasonable way for you to trace it because it was showing "snippet" of your code throwing exception but it was still giving you stack trace from within Alpine. This would usually stop Alpine from continuing to execute other expressions and if your failing x-data was component with code that was duplicated on page (very common with Alpine) it could be hard to trace which particular element/expression was the problem element.
> 
> In 3.5.0+, when the expression is evaluated and fails, it's being moved off main execution thread, so alpine can more or less ignore it and continue. You especially want this behaviour in production environment so perhaps an unimportant component in footer doesn't cause Alpine to stop evaluating everything else on page. You should be able to see log of problem element/expression which provides same information but as in your example, may be slightly harder to interpret. In Chrome DevTools, you can right click on that element now and use "Reveal in Element Panel" to see more context.
> 
> Perhaps you can experiment with @HugoDF's Alpine Dev Tools (https://github.com/alpine-collective/alpinejs-devtools) and see if it helps with debugging. For our dev team, the changes in 3.5.0 made it much easier to debug (often the only way to debug) so please give it a week or two and report back but I think it's working as intended.

I think I understand the benefits of the new error handling update however I'm unsure if the example provides the element logging that you mention, is this just an edgecase or?

I am using the Alpine devtools, which does help a lot!

Somewhat related, perhaps an `x-catch` directive could also be beneficial to catch errors generated in a specific part of the tree, I'll make a new discussion for this idea.

**Reply by SimoTod** [reply]
In dev console with the new version, if you scroll up before the stacktrace you'll see the warning danddanddand
 mentioned. Something like
 ```
Alpine Expression Error: Unexpected identifier
Expression: "JSON.parse({ foo: ' bar's ' })"
 <div x-data=​"JSON.parse({ foo:​ ' bar's ' }​)​">​<!-- view console to see the error --></div>​ 
 ```
 If you right-click on `<div x-data=​"JSON.parse({ foo:​ ' bar's ' }​)​">​<!-- view console to see the error --></div>​ ` you can also select `Reveal in Elements panel`

**Reply by JeremyJaydan** [reply]
> In dev console with the new version, if you scroll up before the stacktrace you'll see the warning danddanddand mentioned. Something like
> 
> ```
> Alpine Expression Error: Unexpected identifier
> Expression: "JSON.parse({ foo: ' bar's ' })"
> <div x-data=​"JSON.parse({ foo:​ ' bar's ' }​)​">​</div>​ 
> ```
> 
> If you right-click on `<div x-data=​"JSON.parse({ foo:​ ' bar's ' }​)​">​<!-- view console to see the error --></div>​ ` you can also select `Reveal in Elements panel`

Oh wow, I had warnings disabled! Darn it lol

Thanks for the helps guys!
